### ISSUE TITLE
Stop re-exposing private APIs in twisted.trial.unittest

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7701]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#7701|
|Type|task|
|Created|2014-11-01 17:36:40Z|
|Branch|https://github.com/twisted/twisted/tree/trial-public-api-7701|

In [#7005](https://github.com/twisted/twisted/issues/7005) adiroban proposed adding various private APIs to `__all__` to satisfy pyflakes.

It's useful to have all those private APIs in one place, for use internally by trial and its unit tests.

But it would be better if they were instead exposed from a new central private module, to avoid any confusion about which are private and which are public trial APIs.

<details><summary>Searchable metadata</summary>

```
trac-id__7701 7701
type__task task
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_trial_public_api_7701 branches/trial-public-api-7701
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1414863400332149 1414863400332149
changetime__1426948831421689 1426948831421689
version__None None
owner__ 
cc__jml cc__adiroiban
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

Do you have any idea where that central private module should be placed?

I am more than happy to work on fixing this... but I have no idea what should I done to have this ticket fixed.

Can you please update the description to be more specific ?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

This module could be something like `twisted.trial._api`.

Note that the motivation behind exporting these things in `twisted.trial.unittest` originally was to avoid having to spread the knowledge about which modules implemented which APIs throughout the rest of the implementation of trial.  This seemed like a good idea because the overall module organization of the trial implementation is fairly ad hoc, inconsistent, complicated, and driven by obscure considerations (eg `_syncrunner` vs `runner`).  If the implementation were organized more coherently into modules, something like `twisted.trial._api` probably wouldn't be as appealing.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to @adiroiban|
|-|-|

OK. I will try to handle this as part of pyflakes cleanup effort. Thanks for the directions.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

(In [43861]) Branching to trial-public-api-7701.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban removed owner|
|-|-|

here is the initial code / cleanup.

I have moved private stuff from trial.unittest to trial._api

I have removed the objects/classes which are used in a single place, excepting their tests.

_clearSuite is never used outside of decorate() (and only once) so I have removed it from the _api.

I did the same with _setWarningRegistryToNone which is only called once in _collectWarnings

I have also removed _collectWarnings as is only called once inside SynchronousTestCase.run()


The _LogObserver class was also never used outside of its module, as only _logObserver is used in 2 other places.

-------

Since this is private stuff I have not added any deprecation handling.

------

For `__all__` variable I have no idea how to handle the PY3 condition. Should it be conditionally extend for PY3 branch or leave as it is?

Ex

```
if not _PY3:
    from twisted.trial._asyncrunner import (
        TestSuite, TestDecorator, decorate)
    __all__.extend([
        'decorate',
        'TestDecorator',
        'TestSuite',
        ])
```

-------

Please take a look at current code and see if changes make sense.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @adiroiban|
|-|-|

Hi adiroiban,  Thanks for working on this.

First, some subjective observations:

 1. Looking over this change, I find myself -0 on the whole thing.  To resolve the issue of importing private names for re-export, I would rather just import the stuff from where it is defined, making it easier to navigate the code directly; there are not that many modules within Trial.
 1. "API" means "application programming interface".  Which means that there ought to be an _application_ somewhere.  It sort of implies "public".  If it's just an interface for programming itself, it's just ... interfaces.  I'd rather call the module `_internals`.
 1. You're wrapping the module docstring at too short of a margin: 60 or so characters rather than the requisite 79.  I think I pointed you at python-docstring-mode in another ticket recently, have a look at that for the way docstrings should be wrapped :).

Now, on to the required stuff:

 1. Adding `PyUnitResultAdapter` and `TestDecorator` to `unittest.__all__` is an unrelated change to this one, and should be done separately.

If you want to just remove the other `__all__` modification and land this, I am OK with that, but I'd rather just have a change which avoids the complexity of the intermediary module, since there are only 3 names in this "internal API".

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

I really don't care how we import stuff in trial I only care about fixing pyflakes warnings :)

I went for t.trial._api based on exarkun's suggestion.
Now if this path did not produced a nice result, no problem. I will redo the work.

I am also +1 on including them directly so if you would approve such a change, I will do that :)

@exarkun are you OK with direct imports?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_9):
> I really don't care how we import stuff in trial I only care about fixing pyflakes warnings :)

OK ;)

> I am also +1 on including them directly so if you would approve such a change, I will do that :)

OK great, let's include them directly.

> @exarkun are you OK with direct imports?

I seriously doubt exarkun will have strong feelings either way, but he is pretty busy with other things right now and probably will not respond to this promptly, so you shouldn't wait for a reply.  (As you can see from the timeline, his last comment was a week ago, and his activity has been a lot sparser lately.)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban removed owner|
|-|-|

Thanks for the review!

For things which were used/imported into a single module I have already made direct imports.

With latest commit I have removed the _api and removed all private import from t.trial.unittest.

I have added PyUnitResultAdapter in `__all__` of  unittest as it looks like a public thing.

pyflakes is happy now with twisted/trial code and so does tox for 2.6 and 2.7 :)

```
$ pyflakes twisted/trial/
# no errors
$ tox -e py26,py27,py33 twisted.trial
  py26: commands succeeded
  py27: commands succeeded
  py33: commands succeeded
  congratulations :)
```

Sending branch for review and to buildbot.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl commented|
|-|-|

I like this. It looks good to me. +1, I'll merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44175]) Merging trial-public-api-7701: Stop re-exposing private APIs in twisted.trial.unittest

Author: adiroiban
Reviewers: glyph, hawkowl
Fixes: [#7701](https://github.com/twisted/twisted/issues/7701)
